### PR TITLE
Try github actions

### PR DESCRIPTION
**Pull request by stanislav-tkach**
## Overview

I have positive experience with Github actions and would like to try them with this project.

### Definition of Done

- [x] There are no TODOs left in the code
- [x] Change is covered by automated [tests](https://github.com/jni-rs/jni-rs/blob/master/CONTRIBUTING.md#tests)
- [x] The [coding guidelines](https://github.com/jni-rs/jni-rs/blob/master/CONTRIBUTING.md#the-code-style) are followed
- [x] Public API has documentation
- [x] This change is not breaking **or** mentioned in the Changelog
- [x] The [continuous integration build](https://www.travis-ci.org/jni-rs/jni-rs) passes

### CODE REVIEW

**Comment by DarkEld3r** [comment]
The results can be seen here: https://github.com/DarkEld3r/jni-rs/actions/runs/155858041

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+          token: ${{ secrets.GITHUB_TOKEN }}
+          args: -- -D warnings
+
+  audit:
+    name: Security audit
+    runs-on: ubuntu-latest
+
+    steps:
+      - uses: actions/checkout@v2
+      - uses: actions-rs/audit-check@v1
```
Comment (dmitry-timofeev) [inline .github/workflows/ci.yml:56]: Interesting that this step takes about 6 minutes :fearful: 

https://github.com/DarkEld3r/jni-rs/runs/832278844?check_suite_focus=true

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+    strategy:
+      matrix:
+        build: [Linux, macOS, Win32-msvc, Win32-gnu, Win64-msvc, Win64-gnu]
+
+        include:
+          - build: Linux
+            os: ubuntu-latest
+          - build: Linux-nightly
+            os: ubuntu-latest
+            rust: nightly
```
Comment (dmitry-timofeev) [inline .github/workflows/ci.yml:73]: Is there anything like "allow_failures" in Travis?

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+          args: --examples --all
+
+      - name: Test
+        uses: actions-rs/cargo@v1
+        with:
+          command: test
+          args: --all
+
+      - name: Test invocation
+        if: matrix.rust == 'nightly'
```
Comment (dmitry-timofeev) [inline .github/workflows/ci.yml:118]: I think it must be the other way around: we test invocation feature on all platforms, but may skip on nightly (though it doesn't seem to make much sense).

Also, it requires setting the `JAVA_HOME` env variable, hence the job currently fails: https://github.com/DarkEld3r/jni-rs/runs/832279070?check_suite_focus=true#step:6:18

File: .github/workflows/ci.yml (hunk 2 above)
Comment (DarkEld3r) [reply .github/workflows/ci.yml:73]: There is `continue-on-error` option, but unfortunately it isn't supported at the job level. Here is an example of a workaround: https://github.com/ozkriff/zemeroth/blob/master/.github/workflows/ci.yml#L85

File: .github/workflows/ci.yml (hunk 1 above)
Comment (DarkEld3r) [reply .github/workflows/ci.yml:56]: Yes, for some reason it is the longest check.

File: .github/workflows/ci.yml (hunk 3 above)
Comment (DarkEld3r) [reply .github/workflows/ci.yml]: https://github.com/actions/setup-java should install Java and set `JAVA_HOME`. I'm going to look at this closely.

**Comment by stanislav-tkach** [comment]
- Java is installed using [this](https://github.com/actions/setup-java) github action.
- On my projects github actions are faster than Travis (and much faster than Appveyor). As far as I can see, the second check is much faster, but it makes sense to compare timing when I am able to fix everything. :upside_down_face: 
- Yes, it is possible to use shell scripts.

**Comment by stanislav-tkach** [comment]
Currently blocked by https://github.com/actions/setup-java/issues/99.

**Comment by stanislav-tkach** [comment]
Finally I managed to get GA working: https://github.com/stanislav-tkach/jni-rs/actions

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -3,119 +3,107 @@ name: CI
...
             os: macOS-latest
-          - build: Win32-msvc
-            os: windows-latest
-            rust: stable-i686-pc-windows-msvc
-            target: i686-pc-windows-msvc
-          - build: Win32-gnu
-            os: windows-latest
-            rust: stable-i686-pc-windows-gnu
-            target: i686-pc-windows-gnu
+            libjvm: /Users/runner/hostedtoolcache/jdk/8.0.265/x64/jre/lib/jli/
```
Comment (dmitry-timofeev) [inline .github/workflows/ci.yml:68]: I see there is an exact version hardcoded. If a JDK is updated (?in the image), will it break our actions? Can we detect the exact path using our `tests_profile` script?

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -3,119 +3,107 @@ name: CI
...
           target: ${{ matrix.target }}
           profile: minimal
           override: true
-
+      - name: Install java
+        uses: actions/setup-java@v1
+        with:
+          java-version: '1.8.0'
+      - name: Add libjvm to LD_LIBRARY_PATH
+        run: echo "::set-env name=LD_LIBRARY_PATH::${{ matrix.libjvm }}"
```
Comment (dmitry-timofeev) [inline .github/workflows/ci.yml:94]: Reminds me of this tweet: https://twitter.com/pchiusano/status/1300243277652856834 :upside_down_face:

File: .github/workflows/ci.yml (hunk 4 above)
Comment (stanislav-tkach) [reply .github/workflows/ci.yml:68]: I don't like the hard-coded path either. :upside_down_face: 

I guess the script can be used, but unfortunately we need a different script for Windows. I will try that approach.

File: .github/workflows/ci.yml (hunk 1 above)
Comment (stanislav-tkach) [reply .github/workflows/ci.yml:52]: It now takes less than a minute. I guess Github does some caching, so initial checks run longer.

**Comment by stanislav-tkach** [comment]
Finally all checks are green: https://github.com/stanislav-tkach/jni-rs/actions/runs/290080253
I have managed to use the `test_profile` script and added the similar script for Windows.

File: .github/workflows/ci.yml [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+        with:
+          toolchain: stable
+          components: clippy
+          profile: minimal
+          override: true
+      - name: Run clippy
+        uses: actions-rs/clippy-check@v1
+        with:
+          token: ${{ secrets.GITHUB_TOKEN }}
+          args: -- -D warnings
```
Comment (dmitry-timofeev) [inline .github/workflows/ci.yml:44]: There seems to be a difference in arguments: 
cargo clippy --all --tests --all-features -- -D warnings

File: .github/workflows/ci.yml [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+        include:
+          - build: Linux
+            os: ubuntu-latest
+          - build: macOS
+            os: macOS-latest
+          - build: Win64-msvc
+            os: windows-latest
+            rust: stable-x86_64-pc-windows-msvc
+            target: x86_64-pc-windows-msvc
+          - build: Win64-gnu
```
Comment (dmitry-timofeev) [inline .github/workflows/ci.yml:71]: Do we need both windows targets?

File: .github/workflows/ci.yml [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+        with:
+          token: ${{ secrets.GITHUB_TOKEN }}
+
+  # Tests.
+  test:
+    name: ${{ matrix.build }}
+    runs-on: ${{ matrix.os }}
+    strategy:
+      matrix:
+        include:
```
Comment (dmitry-timofeev) [inline .github/workflows/ci.yml:62]: We no longer test on nightly with "allow_failures"?

File: .github/workflows/ci.yml [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+      - name: Install java
+        uses: actions/setup-java@v1
+        with:
+          java-version: '1.8.0'
+      - name: Build
+        uses: actions-rs/cargo@v1
+        with:
+          command: build
+          args: --examples --all
+      - name: Test
```
Comment (dmitry-timofeev) [inline .github/workflows/ci.yml:94]: Test default features?

**Review by dmitry-timofeev — COMMENTED** [review]
Generally, looks good. Do these actions for clippy, etc complete? Can we do pass everything we can as CLI flags using these actions? Or do they always lag behind the corresponding commands?

File: .github/workflows/ci.yml (hunk 7 above)
Comment (stanislav-tkach) [reply .github/workflows/ci.yml:71]: I'm not sure. On the one hand, we don't have windows-specific code that should be thoroughly checked. On the other hand, checks for each target take about three minutes, so I think it is not a big deal.

File: .github/workflows/ci.yml (hunk 8 above)
Comment (stanislav-tkach) [reply .github/workflows/ci.yml:62]: Github actions have the `continue-on-error` option, but it completely ignores all errors - the build will be "successful" and the only way to see actual errors is to check the logs. I'm not sure if it is useful in this state.

**Comment by stanislav-tkach** [comment]
> Can we do pass everything we can as CLI flags using these actions? 

Sure. Additional arguments can be specified (using `args`) and are passed to the underlying tool.

File: .github/workflows/ci.yml (hunk 8 above)
Comment (dmitry-timofeev) [reply .github/workflows/ci.yml:62]: I see, let's skip nightly then.

**Review by dmitry-timofeev — APPROVED** [review]
LGTM, thanks! 

Do you plan to update badges/instructions/remove travis config separately?

Shellcheck is also gone :)

File: .github/workflows/ci.yml (hunk 7 above)
Comment (dmitry-timofeev) [reply .github/workflows/ci.yml:71]: I'd consider removing one then, so that there is less stuff.

**Comment by stanislav-tkach** [comment]
Yes, I planned to deal with Travis and badges in a separate pull request.

As for `shellcheck`: I have moved it into a separate CI step, because we now use the same script for Linux and Macos and the latter doesn't have the `shellcheck` tool installed.

**Review by dmitry-timofeev — APPROVED** [review]

**Pull request merged**